File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

dx, dy = (0, 0, 1, -1), (1, -1, 0, 0)
def solution(maps):
    #가중치 없는 그래프 -> BFS
    n, m = len(maps), len(maps[0])
    q = deque()
    q.append((0, 0))
    while len(q) != 0:
        x, y = q.popleft()
        for i in range(4):
            nextX, nextY = x + dx[i], y + dy[i]
            if 0 <= nextX < n and 0 <= nextY < m and maps[nextX][nextY] == 1:                
                maps[nextX][nextY] += maps[x][y]
                if nextX == n-1 and nextY == m-1:
                    return maps[nextX][nextY]
                q.append((nextX, nextY))
            
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
#include<algorithm>
#include<iostream>
#define MAX 100
using namespace std;

int dx[] = {0, 0, 1, -1};
int dy[] = {1, -1, 0, 0};

int solution(vector<vector<int> > maps)
{

    int n = maps.size();
    int m = maps[0].size();
    vector<vector<bool>> visited(n, vector<bool>(m, false));
    queue<pair<int, int>> q;
    q.push({0, 0});
    visited[0][0] = true;
    while(!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();
        
        for(int i = 0; i < 4; i++) {
            int nextX = x + dx[i];
            int nextY = y + dy[i];
            if(nextX < 0 || n <= nextX || nextY < 0 || m <= nextY) continue;
            if(maps[nextX][nextY] == 0 || visited[nextX][nextY]) continue;
            maps[nextX][nextY] += maps[x][y];
            if(nextX == n-1 && nextY == m-1) return maps[nextX][nextY];
        	q.push({nextX, nextY});
            visited[nextX][nextY] = true;
        }
    }
    return -1;
}